### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by tsegismont**
Each commit can be reviewed individually.

Motivation for changes and measured improvements can be found in https://github.com/vert-x3/vertx-micrometer-metrics/issues/184

### CODE REVIEW

File: src/main/java/io/vertx/micrometer/MetricsNaming.java [hunk 1]
```diff
@@ -576,4 +576,51 @@ public MetricsNaming setPoolQueuePending(String poolQueuePending) {
     this.poolQueuePending = poolQueuePending;
     return this;
   }
+
+  public MetricsNaming withBaseName(String baseName) {
```
Comment (vietj) [inline src/main/java/io/vertx/micrometer/MetricsNaming.java:580]: is that called often ? if yes we should consider using a stringbuilder with basename and reuse it for each string we want to produce

File: src/main/java/io/vertx/micrometer/impl/AbstractMetrics.java [hunk 2]
```diff
@@ -70,23 +80,59 @@ public String baseName() {
     return category == null ? null : "vertx." + category + ".";
   }
 
-  Counters counters(String name, String description, Label... keys) {
-    return new Counters(baseName() + name, description, registry, keys);
+  Counter counter(String name, String description, Tags tags) {
+    Meter.Id id = new Meter.Id(name, tags, null, description, COUNTER);
+    Counter c = meterCache.get(id);
```
[the remark this answers is not in the document]
Comment (vietj) [reply src/main/java/io/vertx/micrometer/impl/AbstractMetrics.java:85]: I see the impl is currently using a thread local, but if we make this pluggable for the user, we should have a CAS operation

File: src/main/java/io/vertx/micrometer/impl/VertxEventBusMetrics.java [hunk 3]
```diff
@@ -15,152 +15,152 @@
...
-  private static boolean isValid(Handler handler) {
-    return handler != null && handler.address != null;
-  }
-
-  static class Handler {
-    private final String address;
-
-    Handler(String address) {
-      this.address = address;
+  class Handler {
```
Comment (vietj) [inline src/main/java/io/vertx/micrometer/impl/VertxEventBusMetrics.java:144]: Handlers ?

File: src/main/java/io/vertx/micrometer/MetricsNaming.java (hunk 1 above)
Comment (tsegismont) [reply src/main/java/io/vertx/micrometer/MetricsNaming.java:580]: No, it's called when a metrics object is created (i.e. when Vert.x invokes methods like `io.vertx.core.spi.metrics.VertxMetrics#createHttpServerMetrics`

File: src/main/java/io/vertx/micrometer/impl/AbstractMetrics.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsegismont) [reply src/main/java/io/vertx/micrometer/impl/AbstractMetrics.java:85]: Right, `MeterCache` is thread-safe because it's implemented with a map stored `FastThreadLocal`.

It is not planned to make this pluggable.

[the remark this answers is not in the document]
Comment (vietj) [reply src/main/java/io/vertx/micrometer/impl/AbstractMetrics.java:85]: ok then I think we should simply remove the abstraction and not have a cache that ignores everything and instead check nullity that would simplify the code

File: src/main/java/io/vertx/micrometer/impl/VertxEventBusMetrics.java (hunk 3 above)
Comment (tsegismont) [reply src/main/java/io/vertx/micrometer/impl/VertxEventBusMetrics.java:144]: I didn't change that name. For consistency, I can rename it to `HandlerMetric` (we have `NetServerSocketMetric`, `RequestMetric`, ... etc)

Comment (tsegismont) [reply src/main/java/io/vertx/micrometer/impl/VertxEventBusMetrics.java]: Done, see bfce025

File: src/main/java/io/vertx/micrometer/impl/AbstractMetrics.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsegismont) [reply src/main/java/io/vertx/micrometer/impl/AbstractMetrics.java]: Done

File: src/main/java/io/vertx/micrometer/MetricsNaming.java (hunk 1 above)
Comment (vietj) [reply src/main/java/io/vertx/micrometer/MetricsNaming.java:580]: ah right, that's fine then!

File: src/main/java/io/vertx/micrometer/impl/MeterCache.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+package io.vertx.micrometer.impl;
+
+import io.micrometer.core.instrument.Meter;
+import io.netty.util.concurrent.FastThreadLocal;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class MeterCache {
```
Comment (vietj) [inline src/main/java/io/vertx/micrometer/impl/MeterCache.java:26]: make it package private

File: src/main/java/io/vertx/micrometer/impl/VertxEventBusMetrics.java (hunk 3 above)
Comment (vietj) [reply src/main/java/io/vertx/micrometer/impl/VertxEventBusMetrics.java]: ah right, I thought it was named Handler because it was grouping of bunch of field named handlers, that makes sense then

File: src/main/java/io/vertx/micrometer/impl/MeterCache.java (hunk 4 above)
Comment (tsegismont) [reply src/main/java/io/vertx/micrometer/impl/MeterCache.java]: Done

File: src/main/java/io/vertx/micrometer/impl/MeterCacheImpl.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+import io.micrometer.core.instrument.Meter;
+import io.netty.util.concurrent.FastThreadLocal;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class MeterCacheImpl implements MeterCache {
+
+  private final FastThreadLocal<Map<Meter.Id, Object>> threadLocal;
```
Comment (franz1981) [inline src/main/java/io/vertx/micrometer/impl/MeterCacheImpl.java]: @tsegismont we recently had a problem with @vietj on the CombinerExecutor because of a non static Fast thread local, which was causing some silly leak (which could lead to CVE) due to how it works under the hood.
TLDR make sure not many cache instances could be created in the lifecycle of the application!

Comment (tsegismont) [reply src/main/java/io/vertx/micrometer/impl/MeterCacheImpl.java]: In fact, we talked about this @vietj already :smile:

I replaced `FastThreadLocal` cache with a (less efficient) Vert.x EL Context cache. And then Micrometer team told us about their `MeterProvider` API, which gave good enough results (see the end of the blog post published yesterday).

Anyway, thanks for the heads-up!
